README.MD
# Honepot for CVE-2019-19781 (Citrix ADC)

Detect and log CVE-2019-19781 scan and exploitation attempts.

## Requirements

- python3
- openssl

## Usage

### Step #1: Clone repo

```bash
git clone https://github.com/MalwareTech/CitrixHoneypot.git CitrixHoneypot && cd CitrixHoneypot
```

### Step #2: Generate a self-signed SSL certificate

```bash
openssl req -newkey rsa:2048 -nodes -keyout ssl/key.pem -x509 -days 365 -out ssl/cert.pem
```

### Step #3: Run the honeypot

```bash
nohup python3 -u CitrixHoneypot.py </dev/null >/dev/null 2>ssl/errors.log &
```

## Docker Usage (Optional)

```bash
docker build -t citrixhoneypot .
docker run -d -p 443:443 -v /<insert-homepath>/CitrixHoneypot:/CitrixHoneypot -w /CitrixHoneypot citrixhoneypot
```

## Licencing Agreement: MalwareTech Public Licence

This software is free to use providing the user yells "Oh no, the cyberhackers are coming!" prior to each installation.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [2.0.3]

### Changed in version 2.0.3

* Nothing yet

### Added in version 2.0.3

* Nothing yet

## [2.0.2]

### Changed in version 2.0.2

* Now handling custom HTTP requests, not just HEAD, GET, and POST
* Fixed an error in the MySQL plugin error handler
* Improved the `Dockerfile`
* Minor optimizations

## [2.0.1]

### Changed in version 2.0.1

* Fixed a bug when responding to certain requests

## [2.0.0]

### Added in version 2.0.0

* A script for starting, stopping, and restarting the honeypot
* Config file support
* Various command-line options
* HEAD requests are now logged too
* Output plugin support
* Output plugin for JSON
* Output plugin for MySQL
* Log rotation

### Changed in version 2.0.0

* Made the script compatible with Python 2.7
* The HTTPS server and the logging now use the Twisted framework
* Rewrote the documentation


core/tools.py

from errno import EEXIST
from os import makedirs, path
from datetime import datetime
from socket import socket, AF_INET, SOCK_DGRAM

from core.config import CONFIG

from twisted.python import log

try:
    from urllib.parse import urlsplit, urlunsplit
except ImportError:
    from urlparse import urlsplit, urlunsplit


def get_real_ip (request):
    ip = request.getHeader('X-Real-IP')
    return request.getClientAddress().host if ip is None else ip


def get_real_port (request):
    port = request.getHeader('X-Real-Port')
    return request.getClientAddress().port if port is None else port


def getutctime(unixtime):
    return datetime.utcfromtimestamp(unixtime).isoformat() + 'Z'


def getlocalip():
    s = socket(AF_INET, SOCK_DGRAM)
    try:
        s.connect(('10.255.255.255', 1))
        ip = s.getsockname()[0]
    except:
        ip = '127.0.0.1'
    finally:
        s.close()
    return ip


def resolve_url(url):
    parts = list(urlsplit(url))
    segments = parts[2].split('/')
    segments = [segment + '/' for segment in segments[:-1]] + [segments[-1]]
    resolved = []
    for segment in segments:
        if segment in ('../', '..'):
            if resolved[1:]:
                resolved.pop()
        elif segment not in ('./', '.'):
            resolved.append(segment)
    parts[2] = ''.join(resolved)
    return urlunsplit(parts)


def logger(request, log_level, msg):
    ip = get_real_ip(request)
    port = get_real_port(request)
    log.msg('[{}] ({}:{}): {}'.format(log_level, ip, port, msg))


def write_event(event, cfg):
    output_plugins = cfg['output_plugins']
    for plugin in output_plugins:
        try:
            plugin.write(event)
        except Exception as e:
            log.err(e)
            continue


def mkdir(dir_path):
    if not dir_path:
        return
    try:
        makedirs(dir_path)
    except OSError as exc:
        if exc.errno == EEXIST and path.isdir(dir_path):
            pass
        else:
            raise


def import_plugins(cfg):
    # Load output modules (inspired by the Cowrie honeypot)
    log.msg('Loading the plugins...')
    output_plugins = []
    general_options = cfg
    for x in CONFIG.sections():
        if not x.startswith('output_'):
            continue
        if CONFIG.getboolean(x, 'enabled') is False:
            continue
        engine = x.split('_')[1]
        try:
            output = __import__('output_plugins.{}'.format(engine),
                                globals(), locals(), ['output'], 0).Output(general_options)
            output_plugins.append(output)
            log.msg('Loaded output engine: {}'.format(engine))
        except ImportError as e:
            log.err('Failed to load output engine: {} due to ImportError: {}'.format(engine, e))
        except Exception as e:
            log.err('Failed to load output engine: {} {}'.format(engine, e))
    return output_plugins


def stop_plugins(cfg):
    log.msg('Stoping the plugins...')
    for plugin in cfg['output_plugins']:
        try:
            plugin.stop()
        except Exception as e:
            log.err(e)
            continue



core/config.py

import configparser

from os import environ


def to_environ_key(key):
    return key.upper()


class EnvironmentConfigParser(configparser.ConfigParser):

    def has_option(self, section, option):
        if to_environ_key('_'.join((section, option))) in environ:
            return True
        return super(EnvironmentConfigParser, self).has_option(section, option)

    def get(self, section, option, **kwargs):
        key = to_environ_key('_'.join((section, option)))
        if key in environ:
            return environ[key]
        return super(EnvironmentConfigParser, self).get(section, option, **kwargs)


def readConfigFile(cfgfile):
    """
    Read config files and return ConfigParser object

    @param cfgfile: filename or array of filenames
    @return: ConfigParser object
    """
    parser = EnvironmentConfigParser(interpolation=configparser.ExtendedInterpolation())
    parser.read(cfgfile)
    return parser


CONFIG = readConfigFile(('etc/citrixhoneypot.cfg.base', 'etc/citrixhoneypot.cfg', 'citrixhoneypot.cfg'))


core/logfile.py

import sys

from datetime import datetime

from twisted.python.logfile import DailyLogFile
from twisted.python import log, util


class HoneypotDailyLogFile(DailyLogFile):
    """
    Overload original Twisted with improved date formatting
    """

    def suffix(self, tupledate):
        """
        Return the suffix given a (year, month, day) tuple or unixtime
        """
        try:
            return "{:02d}-{:02d}-{:02d}".format(tupledate[0], tupledate[1], tupledate[2])
        except Exception:
            # try taking a float unixtime
            return '_'.join(map(str, self.toDate(tupledate)))


def myFLOemit(self, eventDict):
    """
    Format the given log event as text and write it to the output file.

    @param eventDict: a log event
    @type eventDict: L{dict} mapping L{str} (native string) to L{object}
    """

    # Custom emit for FileLogObserver
    text = log.textFromEventDict(eventDict)
    if text is None:
        return
    timeStr = self.formatTime(eventDict['time'])
    fmtDict = {
        'text': text.replace('\n', '\n\t')
    }
    msgStr = log._safeFormat('%(text)s\n', fmtDict)
    util.untilConcludes(self.write, timeStr + ' ' + msgStr)
    util.untilConcludes(self.flush)


def myFLOformatTime(self, when):
    """
    Log time in UTC

    By default it's formatted as an ISO8601-like string (ISO8601 date and
    ISO8601 time separated by a space). It can be customized using the
    C{timeFormat} attribute, which will be used as input for the underlying
    L{datetime.datetime.strftime} call.

    @type when: C{int}
    @param when: POSIX (ie, UTC) timestamp.

    @rtype: C{str}
    """
    timeFormatString = self.timeFormat
    if timeFormatString is None:
        timeFormatString = '[%Y-%m-%d %H:%M:%S.%fZ]'
    return datetime.utcfromtimestamp(when).strftime(timeFormatString)


def set_logger(cfg_options):
    log.FileLogObserver.emit = myFLOemit
    log.FileLogObserver.formatTime = myFLOformatTime
    if cfg_options['logfile'] is None:
        log.startLogging(sys.stdout)
    else:
        log.startLogging(HoneypotDailyLogFile.fromFullPath(cfg_options['logfile']), setStdout=False)


core/protocol.py

from time import time

from core import tools

from twisted.python import log
from twisted.web.resource import Resource

try:
    from urllib.parse import unquote, parse_qs
except ImportError:
    from urlparse import unquote, parse_qs


class Index(Resource):
    isLeaf = True
    page_cache = {
        '403.html': '',
        'login.html': '',
        'smb.conf': '',
        'gold_star.html': ''
    }

    def __init__(self, options):
        self.cfg = options

    def render_HEAD(self, request):
        path = unquote(request.uri.decode('utf-8'))
        method = request.method.decode('utf-8')

        tools.logger(request, 'INFO', '{}: {}'.format(method, path))

        # split the path by '/', ignoring empty string
        url_path = list(filter(None, path.split('/')))

        if path.find('/../') != -1:
            # flatten path to ease parsing
            collapsed_path = tools.resolve_url(path)
            url_path = list(filter(None, collapsed_path.split('/')))

            # check if the directory traversal bug has been tried
            if len(url_path) >= 1 and url_path[0] == 'vpns':

                unix_time = time()
                human_time = tools.getutctime(unix_time)
                local_ip = tools.getlocalip()
                event = {
                    'eventid': 'citrix.connection',
                    'timestamp': human_time,
                    'unixtime': unix_time,
                    'src_ip': tools.get_real_ip(request),
                    'src_port': tools.get_real_port(request),
                    'dst_ip': local_ip,
                    'dst_port': self.cfg['port'],
                    'sensor': self.cfg['sensor'],
                    'request': method,
                    'url': path
                }

                # 403 on /vpn/../vpns/ is used by some scanners to detect vulnerable hosts
                # Ex: https://github.com/cisagov/check-cve-2019-19781/blob/develop/src/check_cve/check.py
                if len(url_path) == 1 and url_path[0] == 'vpns':
                    tools.logger(request, 'WARNING', 'Detected type 1 CVE-2019-19781 scan attempt!')
                    event['message'] = 'Scan type 1'

                # some scanners try to fetch smb.conf to detect vulnerable hosts
                # Ex: https://github.com/trustedsec/cve-2019-19781/blob/master/cve-2019-19781_scanner.py
                elif collapsed_path == '/vpns/cfg/smb.conf':
                    tools.logger(request, 'WARNING', 'Detected type 2 CVE-2019-19781 scan attempt!')
                    event['message'] = 'Scan type 2'

                # some scanners try to fetch services.html to detect vulnerable hosts
                # Ex: https://github.com/mekoko/CVE-2019-19781/blob/master/CVE-2019-19781.py
                elif collapsed_path == '/vpns/services.html':
                    tools.logger(request, 'WARNING', 'Detected type 3 CVE-2019-19781 scan attempt!')
                    event['message'] = 'Scan type 3'

                # we got a request that sort of matches CVE-2019-19781, but it's not a known scan attempt
                else:
                    tools.logger(request, 'DEBUG', 'Error: unhandled CVE-2019-19781 scan attempt: {}'.format(path))
                    event['message'] = 'Unknown scan'
                tools.write_event(event, self.cfg)

        return self.send_response(request)

    def render_GET(self, request):
        path = unquote(request.uri.decode('utf-8'))
        method = request.method.decode('utf-8')

        tools.logger(request, 'INFO', '{}: {}'.format(method, path))

        if self.struggle_check(request, path):
            self.send_response(request)

        # split the path by '/', ignoring empty string
        url_path = list(filter(None, path.split('/')))

        # if url is empty or path is /vpn/, display fake login page
        if len(url_path) == 0 or \
           (len(url_path) == 1 and url_path[0] == 'vpn') or \
           (len(url_path) == 2 and url_path[0] == 'vpn' and url_path[1].lower().startswith('index.htm')):
            return self.send_response(request, self.get_page('login.html'))

        # only proceed if a directory traversal was attempted
        if path.find('/../') != -1:
            # flatten path to ease parsing
            collapsed_path = tools.resolve_url(path)
            url_path = list(filter(None, collapsed_path.split('/')))

            # check if the directory traversal bug has been tried
            if len(url_path) >= 1 and url_path[0] == 'vpns':

                unix_time = time()
                human_time = tools.getutctime(unix_time)
                local_ip = tools.getlocalip()
                event = {
                    'eventid': 'citrix.connection',
                    'timestamp': human_time,
                    'unixtime': unix_time,
                    'src_ip': tools.get_real_ip(request),
                    'src_port': tools.get_real_port(request),
                    'dst_ip': local_ip,
                    'dst_port': self.cfg['port'],
                    'sensor': self.cfg['sensor'],
                    'request': method,
                    'url': path
                }

                # 403 on /vpn/../vpns/ is used by some scanners to detect vulnerable hosts
                # Ex: https://github.com/cisagov/check-cve-2019-19781/blob/develop/src/check_cve/check.py
                if len(url_path) == 1 and url_path[0] == 'vpns':
                    tools.logger(request, 'WARNING', 'Detected type 1 CVE-2019-19781 scan attempt!')
                    event['message'] = 'Scan type 1'
                    response = self.get_page('403.html').replace('{url}', collapsed_path)

                # some scanners try to fetch smb.conf to detect vulnerable hosts
                # Ex: https://github.com/trustedsec/cve-2019-19781/blob/master/cve-2019-19781_scanner.py
                elif collapsed_path == '/vpns/cfg/smb.conf':
                    tools.logger(request, 'WARNING', 'Detected type 2 CVE-2019-19781 scan attempt!')
                    event['message'] = 'Scan type 2'
                    response = self.get_page('smb.conf')

                # some scanners try to fetch services.html to detect vulnerable hosts
                # Ex: https://github.com/mekoko/CVE-2019-19781/blob/master/CVE-2019-19781.py
                elif collapsed_path == '/vpns/services.html':
                    tools.logger(request, 'WARNINg', 'Detected type 3 CVE-2019-19781 scan attempt!')
                    event['message'] = 'Scan type 3'
                    response = self.get_page('smb.conf')

                elif len(url_path) >= 2 and url_path[0] == 'vpns' and url_path[1] == 'portal':
                    tools.logger(request, 'CRITICAL', 'Detected CVE-2019-19781 completion!')
                    event['message'] = 'Exploit completion'
                    response = ''

                # we got a request that sort of matches CVE-2019-19781, but it's not a known scan attempt
                else:
                    tools.logger(request, 'DEBUG', 'Error: unhandled CVE-2019-19781 scan attempt: {}'.format(path))
                    event['message'] = 'Unknown scan'
                    response = ''
                tools.write_event(event, self.cfg)
                return self.send_response(request, response)

        # if all else fails return nothing
        return self.send_response(request)

    def render_POST(self, request):
        path = unquote(request.uri.decode('utf-8'))

        tools.logger(request, 'INFO', '{}: {}'.format(request.method.decode('utf-8'), path))

        if request.getHeader('Content-Length'):
            collapsed_path = tools.resolve_url(path)
            content_length = int(request.getHeader('Content-Length'))
            if content_length > 0:
                post_data = request.content.read().decode('utf-8')
                unix_time = time()
                tools.logger(request, 'INFO', 'POST body: {}'.format(post_data))
                human_time = tools.getutctime(unix_time)
                local_ip = tools.getlocalip()
                payload = parse_qs(post_data)['title'][0]
                event = {
                    'eventid': 'citrix.payload',
                    'timestamp': human_time,
                    'unixtime': unix_time,
                    'src_ip': tools.get_real_ip(request),
                    'src_port': tools.get_real_port(request),
                    'dst_ip': local_ip,
                    'dst_port': self.cfg['port'],
                    'sensor': self.cfg['sensor'],
                    'request': 'POST',
                    'message': 'Exploit',
                    'body': post_data,
                    'payload': payload,
                    'url': path
                }

                # RCE path is /vpns/portal/scripts/newbm.pl and payload is contained in POST data
                if collapsed_path in ['/vpns/portal/scripts/newbm.pl', '/vpns/portal/scripts/rmbm.pl']:
                    tools.logger(request, 'CRITICAL', 'Detected CVE-2019-19781 payload: {}'.format(payload))

                tools.write_event(event, self.cfg)

        self.struggle_check(request, path)

        # send empty response as we're now done
        return self.send_response(request)

    def render(self, request):
        return self.render_GET(request)

    def struggle_check(self, request, path):
        if self.cfg['struggle']:
            # if the path does not contain /../ it's likely attacker was using a sanitized client which removed it
            if path in ['/vpns/portal/scripts/newbm.pl', '/vpns/cfg/smb.conf', '/vpns/']:
                tools.logger(request, 'DEBUG', 'Detected a failed directory traversal attempt.')
                self.send_response(request, self.get_page('gold_star.html'))
                return True

        return False

    # a simple wrapper to cache files from "responses" folder
    def get_page(self, page):
        # if page is not in cache, load it from file
        if self.page_cache[page] == '':
            with open('responses/{}'.format(page), 'r') as f:
                self.page_cache[page] = f.read()

        return self.page_cache[page]

    # overload base class's send_response() to set appropriate headers and server version
    def send_response(self, request, page=''):
        request.setHeader('Server', 'Apache')
        request.setHeader('Set-Cookie', 'NSC_AAAC=xyz;Path=/;expires=Wednesday, 09-Nov-1999 23:12:40 GMT;Secure')
        request.setHeader('Set-Cookie', 'NSC_EPAC=xyz;Path=/;expires=Wednesday, 09-Nov-1999 23:12:40 GMT;Secure')
        request.setHeader('Set-Cookie', 'NSC_USER=xyz;Path=/;expires=Wednesday, 09-Nov-1999 23:12:40 GMT;Secure')
        request.setHeader('Set-Cookie', 'NSC_TEMP=xyz;Path=/;expires=Wednesday, 09-Nov-1999 23:12:40 GMT;Secure')
        request.setHeader('Set-Cookie', 'NSC_PERS=xyz;Path=/;expires=Wednesday, 09-Nov-1999 23:12:40 GMT;Secure')
        request.setHeader('Set-Cookie', 'NSC_BASEURL=xyz;Path=/;expires=Wednesday, 09-Nov-1999 23:12:40 GMT;Secure')
        request.setHeader('Set-Cookie', 'CsrfToken=xyz;Path=/;expires=Wednesday, 09-Nov-1999 23:12:40 GMT;Secure')
        request.setHeader('Set-Cookie', 'CtxsAuthId=xyz;Path=/;expires=Wednesday, 09-Nov-1999 23:12:40 GMT;Secure')
        request.setHeader('Set-Cookie', 'ASP.NET_SessionId=xyz;Path=/;expires=Wednesday, 09-Nov-1999 23:12:40 GMT;Secure')
        request.setHeader('Set-Cookie', 'NSC_TMAA=xyz;Path=/;expires=Wednesday, 09-Nov-1999 23:12:40 GMT')
        request.setHeader('Set-Cookie', 'NSC_TMAS=xyz;Path=/;expires=Wednesday, 09-Nov-1999 23:12:40 GMT;Secure')
        request.setHeader('Set-Cookie', 'NSC_TEMP=xyz;Path=/;expires=Wednesday, 09-Nov-1999 23:12:40 GMT')
        request.setHeader('Set-Cookie', 'NSC_PERS=xyz;Path=/;expires=Wednesday, 09-Nov-1999 23:12:40 GMT')
        request.setHeader('Connection', 'Close')
        request.setHeader('Content-Length', str(len(page)))
        request.setHeader('Cache-control', 'no-cache, no-store')
        request.setHeader('Pragma', 'no-cache')
        request.setHeader('Content-type', 'text/html')
        return '{}'.format(page).encode('utf-8')


output_plugins/mysql.py

try:
    from MySQLdb import Error, OperationalError
except ImportError:
    try:
        from MySQLdb._exceptions import Error, OperationalError
    except ImportError:
        from _mysql_exceptions import Error, OperationalError

from core import output
from core.config import CONFIG

from sys import exc_info
from hashlib import sha256
from geoip2.database import Reader

from twisted.python import log
from twisted.python.compat import reraise
from twisted.enterprise.adbapi import ConnectionPool


class ReconnectingConnectionPool(ConnectionPool):
    """
    Reconnecting adbapi connection pool for MySQL.
    This class improves on the solution posted at
    http://www.gelens.org/2008/09/12/reinitializing-twisted-connectionpool/
    by checking exceptions by error code and only disconnecting the current
    connection instead of all of them.
    Also see:
    http://twistedmatrix.com/pipermail/twisted-python/2009-July/020007.html
    """

    def _runInteraction(self, interaction, *args, **kw):

        def rerise_exception(conn):
            _, excValue, excTraceback = exc_info()
            try:
                conn.rollback()
            except:
                log.msg('Rollback failed')
            reraise(excValue, excTraceback)

        conn = self.connectionFactory(self)
        trans = self.transactionFactory(self, conn)
        try:
            result = interaction(trans, *args, **kw)
            trans.close()
            conn.commit()
            return result
        except OperationalError as e:
            if e.args[0] not in (2003, 2006, 2013):
                rerise_exception(conn)
            else:
                conn = self.connections.get(self.threadID())
                self.disconnect(conn)
                return ConnectionPool._runInteraction(self, interaction, *args, **kw)
        except:
            rerise_exception(conn)


class Output(output.Output):

    def __init__(self, general_options):
        self.host = CONFIG.get('output_mysql', 'host', fallback='localhost')
        self.database = CONFIG.get('output_mysql', 'database', fallback='')
        self.user = CONFIG.get('output_mysql', 'username', fallback='')
        self.password = CONFIG.get('output_mysql', 'password', fallback='', raw=True)
        self.port = CONFIG.getint('output_mysql', 'port', fallback=3306)

        self.geoipdb_city_path = CONFIG.get('output_mysql', 'geoip_citydb', fallback='')
        self.geoipdb_asn_path = CONFIG.get('output_mysql', 'geoip_asndb', fallback='')

        self.debug = CONFIG.getboolean('output_mysql', 'debug', fallback=False)
        self.geoip = CONFIG.getboolean('output_mysql', 'geoip', fallback=True)

        self.dbh = None

        output.Output.__init__(self, general_options)

    def local_log(self, msg):
        if self.debug:
            log.msg(msg)

    def start(self):
        try:
            self.dbh = ReconnectingConnectionPool(
                'MySQLdb',
                host=self.host,
                db=self.database,
                user=self.user,
                passwd=self.password,
                port=self.port,
                charset='utf8',
                use_unicode=True,
                cp_min=1,
                cp_max=1
            )
        except Error as e:
            self.local_log('MySQL plugin: Error {}: {}'.format(e.args[0], e.args[1]))

        if self.geoip:
            try:
                self.reader_city = Reader(self.geoipdb_city_path)
            except:
                self.local_log('Failed to open City GeoIP database {}'.format(self.geoipdb_city_path))

            try:
                self.reader_asn = Reader(self.geoipdb_asn_path)
            except:
                self.local_log('Failed to open ASN GeoIP database {}'.format(self.geoipdb_asn_path))

    def stop(self):
        if self.geoip:
            if self.reader_city is not None:
               self.reader_city.close()
            if self.reader_asn is not None:
               self.reader_asn.close()

    def write(self, event):
        """
        TODO: Check if the type (date, datetime or timestamp) of columns is appropriate for your needs and timezone
        - MySQL Documentation - The DATE, DATETIME, and TIMESTAMP Types
            (https://dev.mysql.com/doc/refman/5.7/en/datetime.html):
        "MySQL converts TIMESTAMP values from the current time zone to UTC for storage,
        and back from UTC to the current time zone for retrieval.
        (This does not occur for other types such as DATETIME.)"
        """
        if 'connection' in event['eventid']:
            self.dbh.runInteraction(self.connect_event, event, False)
        elif 'payload' in event['eventid']:
            self.dbh.runInteraction(self.connect_event, event, True)

    def simple_query(self, txn, sql, args):
        if self.debug:
            if len(args):
                self.local_log("output_mysql: MySQL query: {} {}".format(sql, repr(args)))
            else:
                self.local_log("output_mysql: MySQL query: {}".format(sql))
        try:
            if len(args):
                txn.execute(sql, args)
            else:
                txn.execute(sql)
            result = txn.fetchall()
        except Exception as e:
            self.local_log('output_mysql: MySQL Error: {}'.format(e))
            result = None
        return result

    def get_id(self, txn, table, column, entry):
        r = self.simple_query(txn, "SELECT `id` FROM `{}` WHERE `{}` = %s".format(table, column), (entry, ))
        if r:
            id = r[0][0]
        else:
            self.simple_query(txn, "INSERT INTO `{}` (`{}`) VALUES (%s)".format(table, column), (entry, ))
            r = self.simple_query(txn, 'SELECT LAST_INSERT_ID()', ())
            if r:
                id = int(r[0][0])
            else:
                id = 0
        return id

    def get_hashed_id(self, txn, table, entry):
        sc = entry.strip()
        shasum = sha256(sc).hexdigest()
        r = self.simple_query(txn, "SELECT `id` FROM `{}` WHERE `inputhash` = %s".format(table), (shasum, ))
        if r:
            id = int(r[0][0])
        else:
            self.simple_query(txn, "INSERT INTO `{}` (`input`, `inputhash`) VALUES (%s, %s)".format(table),
                              (sc.decode('utf-8').encode('unicode_escape'), shasum, ))
            r = self.simple_query(txn, 'SELECT LAST_INSERT_ID()', ())
            if r:
                id = int(r[0][0])
            else:
                id = 0
        return id

    def geolocate(self, remote_ip):
        try:
            response_city = self.reader_city.city(remote_ip)
            city = response_city.city.name
            if city is None:
                city = ''
            country = response_city.country.name
            if country is None:
                country = ''
                country_code = ''
            else:
                country_code = response_city.country.iso_code
        except Exception as e:
            self.local_log(e)
            city = ''
            country = ''
            country_code = ''

        try:
            response_asn = self.reader_asn.asn(remote_ip)
            if response_asn.autonomous_system_organization is not None:
                org = response_asn.autonomous_system_organization.encode('utf8')
            else:
                org = ''

            if response_asn.autonomous_system_number is not None:
                asn_num = response_asn.autonomous_system_number
            else:
                asn_num = 0
        except Exception as e:
            self.local_log(e)
            org = ''
            asn_num = 0
        return country, country_code, city, org, asn_num

    def connect_event(self, txn, event, has_payload):
        remote_ip = event['src_ip']

        path_id = self.get_id(txn, 'paths', 'path', event['url'])
        message_id = self.get_id(txn, 'messages', 'message', event['message'])
        sensor_id = self.get_id(txn, 'sensors', 'name', event['sensor'])

        if has_payload:
            body_id = self.get_hashed_id(txn, 'bodies', event['body'])
            payload_id = self.get_hashed_id(txn, 'payloads', event['payload'])
            self.simple_query(txn, """
                INSERT INTO `connections` (
                    `timestamp`, `ip`, `remote_port`, `request`, `path`, `body`,
                    `payload`, `message`, `local_host`, `local_port`, `sensor`)
                VALUES (FROM_UNIXTIME(%s), %s, %s, %s, %s, %s, %s, %s, %s, %s, %s)
                """,
                (event['unixtime'], remote_ip, event['src_port'], event['request'], path_id,
                body_id, payload_id, message_id, event['dst_ip'], event['dst_port'], sensor_id, ))
        else:
            self.simple_query(txn, """
                INSERT INTO `connections` (
                    `timestamp`, `ip`, `remote_port`, `request`, `path`,
                    `message`, `local_host`, `local_port`, `sensor`)
                VALUES (FROM_UNIXTIME(%s), %s, %s, %s, %s, %s, %s, %s, %s)
                """,
                (event['unixtime'], remote_ip, event['src_port'], event['request'],
                path_id, message_id, event['dst_ip'], event['dst_port'], sensor_id, ))

        if self.geoip:
            country, country_code, city, org, asn_num = self.geolocate(remote_ip)
            self.simple_query(txn, """
                INSERT INTO `geolocation` (`ip`, `country_name`, `country_iso_code`, `city_name`, `org`, `org_asn`)
                VALUES (%s, %s, %s, %s, %s, %s)
                ON DUPLICATE KEY UPDATE
                    `country_name` = %s,
                    `country_iso_code` = %s,
                    `city_name` = %s,
                    `org` = %s,
                    `org_asn` = %s
                """,
                (remote_ip, country, country_code, city, org, asn_num, country, country_code, city, org, asn_num, ))


output_plugins/jsonlog.py

import os

from json import dump
from errno import EEXIST
from copy import deepcopy

from core import output
from core.config import CONFIG
from core.logfile import HoneypotDailyLogFile

class Output(output.Output):

    def start(self):
        self.epoch_timestamp = CONFIG.getboolean('output_jsonlog', 'epoch_timestamp', fallback=False)
        fn = CONFIG.get('output_jsonlog', 'logfile')
        dirs = os.path.dirname(fn)
        base = os.path.basename(fn)
        if not os.path.exists(dirs) and os.sep in fn:
            try:
                os.makedirs(dirs)
            except OSError as exc:
                if exc.errno != EEXIST:
                    raise
        self.outfile = HoneypotDailyLogFile(base, dirs, defaultMode=0o664)

    def stop(self):
        self.outfile.flush()

    def write(self, event):
        if not self.epoch_timestamp:
            # We need 'unixtime' value in some other plugins
            event_dump = deepcopy(event)
            event_dump.pop('unixtime', None)
        else:
            event_dump = event
        dump(event_dump, self.outfile, separators=(',', ':'))
        self.outfile.write('\n')
        self.outfile.flush()


CitrixHoneypot.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""
Licencing Agreement: MalwareTech Public Licence
This software is free to use providing the user yells
"Oh no, the cyberhackers are coming!" prior to each installation.
"""

from os.path import join
from socket import gethostname
from argparse import ArgumentParser

from core.config import CONFIG
from core.protocol import Index
from core.logfile import set_logger
from core.tools import mkdir, import_plugins, stop_plugins

from twisted.web import server
from twisted.python import log
from twisted.internet import reactor, endpoints


__VERSION__ = '2.0.3'
__description__ = 'Citrix CVE-2019-19781 Honeypot by MalwareTech'


def get_options(cfg_options):
    parser = ArgumentParser(description=__description__)

    parser.add_argument('-v', '--version', action='version', version='%(prog)s version ' + __VERSION__)
    parser.add_argument('-a', '--addr', type=str, default=cfg_options['addr'],
                        help='Address to bind to (default: {})'.format(cfg_options['addr']))
    parser.add_argument('-p', '--port', type=int, default=cfg_options['port'],
                        help='Port to listen on (default: {})'.format(cfg_options['port']))
    parser.add_argument('-l', '--logfile', type=str, default=cfg_options['logfile'],
                        help='Log file (default: stdout)')
    parser.add_argument('-d', '--ssldir', type=str, default=cfg_options['ssldir'],
                        help='Directory of the SSL certificate (default: {})'.format(cfg_options['ssldir']))
    parser.add_argument('-s', '--sensor', type=str, default=cfg_options['sensor'],
                        help='Sensor name (default: {})'.format(cfg_options['sensor']))

    args = parser.parse_args()
    return args


def mySiteLog(request):
    """
    Empty log formatter to suppress the normal logging of
    the web requests, since we'll be doing our own logging.
    """
    return


def main():
    cfg_options = {}
    cfg_options['addr'] = CONFIG.get('honeypot', 'out_addr', fallback='0.0.0.0')
    cfg_options['port'] = CONFIG.getint('honeypot', 'listen_port', fallback=443)
    log_name = CONFIG.get('honeypot', 'log_filename', fallback='')
    if log_name:
        logdir = CONFIG.get('honeypot', 'log_path', fallback='')
        mkdir(logdir)
        cfg_options['logfile'] = join(logdir, log_name)
    else:
        cfg_options['logfile'] = None
    cfg_options['ssldir'] = CONFIG.get('honeypot', 'ssl_dir', fallback='ssl')
    cfg_options['sensor'] = CONFIG.get('honeypot', 'sensor_name', fallback=gethostname())
    cfg_options['debug'] = CONFIG.get('honeypot', 'verbosity', fallback='info')
    cfg_options['struggle'] = CONFIG.getboolean('honeypot', 'struggle_check', fallback=False)

    args = get_options(cfg_options)

    cfg_options['addr'] = args.addr
    cfg_options['port'] = args.port
    cfg_options['logfile'] = args.logfile
    cfg_options['ssldir'] = args.ssldir
    cfg_options['sensor'] = args.sensor

    set_logger(cfg_options)

    log.msg(__description__)

    cfg_options['output_plugins'] = import_plugins(cfg_options)

    site = server.Site(Index(cfg_options))
    site.log = mySiteLog
    endpoint_spec = 'ssl:interface={}:port={}:privateKey={}/key.pem:certKey={}/cert.pem'.format(
        cfg_options['addr'],
        cfg_options['port'],
        cfg_options['ssldir'],
        cfg_options['ssldir']
    )
    log.msg('Listening on {}:{}.'.format(cfg_options['addr'], cfg_options['port']))
    endpoints.serverFromString(reactor, endpoint_spec).listen(site)
    reactor.run()   # pylint: disable=no-member
    log.msg('Shutdown requested, exiting...')
    stop_plugins(cfg_options)


if __name__ == '__main__':
    main()
